File: striver-sde-sheet/432-767-Reorganize-String/solution.js
/**
 * @param {string} s
 * @return {string}
 */
var reorganizeString = function(s) {
    const count = {};
    for (let char of s) {
        count[char] = (count[char] || 0) + 1;
    }
    
    const maxCount = Math.max(...Object.values(count));
    if (maxCount > Math.floor((s.length + 1) / 2)) {
        return "";
    }
    
    const sorted = Object.keys(count).sort((a, b) => count[b] - count[a]);
    const result = [];
    let prev = null;
    
    while (sorted.length > 0 || prev) {
        if (sorted.length === 0 && prev) {
            return "";
        }
        
        const curr = sorted[0];
        result.push(curr);
        count[curr]--;
        
        if (count[curr] === 0) {
            sorted.shift();
        } else {
            sorted.sort((a, b) => count[b] - count[a]);
        }
        
        if (prev) {
            sorted.push(prev);
            sorted.sort((a, b) => count[b] - count[a]);
            prev = null;
        }
        
        if (count[curr] > 0) {
            prev = curr;
        }
    }
    
    return result.join("");
};

File: striver-sde-sheet/432-767-Reorganize-String/solution.py
from collections import Counter
import heapq

class Solution:
    def reorganizeString(self, s: str) -> str:
        count = Counter(s)
        max_count = max(count.values())
        if max_count > (len(s) + 1) // 2:
            return ""
        
        heap = [(-freq, char) for char, freq in count.items()]
        heapq.heapify(heap)
        
        result = []
        prev = None
        
        while heap:
            freq, char = heapq.heappop(heap)
            result.append(char)
            
            if prev and prev[0] < 0:
                heapq.heappush(heap, prev)
            
            prev = (freq + 1, char) if freq + 1 < 0 else None
        
        return "".join(result)
